File: python/worker.py
"""Worker entrypoint (slide 34): polls the task queue, runs workflow + activities.

    uv run worker.py

Kill it mid-conversation and restart it — the loop resumes exactly where it
was. That's the crash-recovery demo beat.
"""

import asyncio
from concurrent.futures import ThreadPoolExecutor

from temporalio.worker import Worker

import config
from activities import control
from activities.llm import call_llm
from activities.tools import execute_tool
from workflows.agent import SupportAgentWorkflow


async def main() -> None:
    client = await config.temporal_client()
    control.set_client(client)  # so call_llm can read its conversation's kill-switch
    with ThreadPoolExecutor(max_workers=8) as activity_executor:
        worker = Worker(
            client,
            task_queue=config.TASK_QUEUE,
            workflows=[SupportAgentWorkflow],
            activities=[call_llm, execute_tool],
            activity_executor=activity_executor,
        )
        print(f"worker polling task queue '{config.TASK_QUEUE}' "
              f"on {config.TEMPORAL_ADDRESS} (provider: {config.LLM_PROVIDER})")
        await worker.run()


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/src/worker.ts
// Worker entrypoint: polls the task queue, runs the workflow + activities (LLM,
// tools, DB, research). Same crash-recovery beat as python/worker.py — kill it
// mid-conversation and restart; the loop resumes from history.
//
//     npm run worker    (or: make worker-ts)

import { Worker } from '@temporalio/worker';
import * as activities from './activities';
import * as control from './activities/control';
import * as config from './config';

async function main(): Promise<void> {
  const connection = await config.workerConnection();
  // A client so LLM activities can read their conversation's kill-switch.
  control.setClient(await config.makeClient());

  const worker = await Worker.create({
    connection,
    namespace: config.TEMPORAL_NAMESPACE,
    taskQueue: config.TASK_QUEUE,
    workflowsPath: require.resolve('./agent'),
    activities,
  });

  console.log(
    `ts worker polling task queue '${config.TASK_QUEUE}' on ${config.TEMPORAL_ADDRESS} ` +
      `(namespace: ${config.TEMPORAL_NAMESPACE}, provider: ${config.LLM_PROVIDER})`
  );
  await worker.run();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
